(CDAP-16353) Make ProperitesResolver use PreferencesFetcher to get properties

wyzhang opened the pull request:
Motivation:
At the moment, ProperitesResolver directly accesses local preferences store.
Change that to use PreferencesFetcher allows us to fetch properties either from
local storage or from a remote service via REST calls. This paves the path for
support per service local level DB.

Additional changes:
- Make PreferencesHttpHandler to return empty properties instead of fail if the
  requested entity (e.g. program, or application) doesn't exist, for 2 reasons
  1) consistent with the behavior of preferences store
  2) user may request resolved properties for a program, which will include
     properties from parents (e.g. application, namespace and instance). Even
     through the program doesn't exist, it would be good to return resolved
     properties from its parent and above.

yaojiefeng commented on cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/DefaultPreviewRunnerModule.java:
```diff
@@ -162,6 +164,8 @@ protected void configure() {
 
     bind(PreviewRequest.class).toInstance(previewRequest);
+
+    bind(PreferencesFetcher.class).to(LocalPreferencesFetcherInternal.class).in(Scopes.SINGLETON);
```
I think preview should use remote?

wyzhang replied:
That change comes in the next PR :), this PR ensure there is no behavior change.

The plan is 
if cloudsql => bind to local, so same as today
if nosql => bind to remote

yaojiefeng commented on cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/PreferencesHttpHandlerInternal.java:
```diff
@@ -77,9 +72,7 @@ public void getNamespacePreferences(HttpRequest request, HttpResponder responder
-    if (!namespaceQueryAdmin.exists(namespaceId)) {
-      throw new NamespaceNotFoundException(namespaceId);
-    }
+    // No need to check if namespace exists. PreferencesService returns an empty PreferencesDetail when that happens.
```
Better not changing the behavior, these existence check is fairly common in our http handler. The pref store is only responsible for fetch/put data, it is handler's responsibility to validate the existence of entity. If an entity doesn't exist, it is no point to return the preferences since we delete the preferences whenever a program is deleted.

wyzhang replied:
Context is: 
This preferences http handler was recently introduced to support per-service local levelDB and it is currently not used anywhere. So changing the behavior will not cause any regression. 

As for validating entity existence, in general I agree with failing if entity doesn't exist. But I think it might make sense for preferences for 2 reasons:
1) preferences support "resolve", so if the app doesn't exist, caller might want get resolved properties from parents. 
2) preview deploys apps locally and fetch *resolved* preferences before run the preview. I think that fetch returns properties from parent as the app is not deployed to appfabric. To achieve the same effect, remote preferences fetch have to handle the entity non-existence case

albertshau replied:
I agree that the remote and local implementations should have the same behavior. Failing on namespace not exists would actually be the backwards incompatible behavior in this case. Have we made sure the other internal endpoints would behave the same as their local counterparts?

Can you add some javadocs for the method that explain the contract? Otherwise it is very easy to look at this class and think it should be checking for existence like all the other handlers.

wyzhang replied:
Done. 

Added javadoc to call it out that request won't fail in the case of non-existing entity. 

Let me refactor the test for local and remote in the next PR to have a parameterized test to ensure local vs remote goes through exact same test cases to validate the same behavior (i.e. similar to what I did in https://github.com/cdapio/cdap/pull/12007 for testing local vs remote ScheduleFetcher)

yaojiefeng replied:
@albertshau the problem is in local http handler we do check the existence of the entity, it is not in `PreferencesService` but in the actually handler. I don't think we should remove it here since it will cause confusion in the future in case there is some step involving getting preferences. 
Preview actually deploys app first and then call the run program method, I believe we get preferences when we actually run the program, at this step, all the entities should already be there. If the check of existence fails there, it means a bug in the app deploy or preview.

yaojiefeng replied:
Ah, I see the changes in `PropertiesResolver`, seems like this is only used there? Then it is fine to remove the existence check to keep the functionality same. Please ignore my previous comment

wyzhang replied:
Yes. PreferencesFetcher is used in propertiesResolver only at the moment. But metadata service will use it as well in the future (to switch from local store access to REST calls)

@yaojiefeng "in local http handler we do check the existence of the entity", I think you are referring to public http handler for preference. Also note that public endpoints return property map while internal ones return PreferencesDetail.

yaojiefeng reviewed (commented):
two comments

wyzhang commented:
Discarding this PR for now.

Include the change in new PR 
https://github.com/cdapio/cdap/pull/12006

albertshau reviewed (approved):
comment about javadocs, otherwise lgtm

wyzhang committed:
(CDAP-16353) Make ProperitesResolver using PreferencesFetcher to get preference properites.

Motivation:
At the moment, ProperitesResolver directly accesses local preferences store.
Change that to use PreferencesFetcher allows us to fetch properties either from
local storage or from a remote service via REST calls. This paves the path for
support per service local level DB.

Additional changes:
- Make PreferencesHttpHandler to return empty properties instead of fail if the
  requested entity (e.g. program, or applciation) doesn't exist, for 2 reasons
  1) consistent with the behavior of preferences store
  2) user may request resolved properties for a program, which will include
     properites from parents (e.g. application, namespace and instance). Even
     through the program doesn't exist, it would be good to return resolved
     properties from its parent and above.
